### PR TITLE
[10.x] Feature: Create non-primary autoincrements in migrations

### PR DESCRIPTION
**Pull request by olliescase**
# The Problem
There is currently no easy and reliable way to add auto-increments into a migration without it being a primary key.

There is debate as to wether this should be desirable, however, I believe it should, at least, be possible. A valid usecase is wanting non-incrementing ids to avoid record number 4563 implicating the existance of 4562, 4561, etc., allowing for easy scraping of data in URLs. Instead, you may wish for your increments to be in a separate column to that which you use to primarily identify the record.

Currently, the only way to achieve this is with something as horrible as:

```php
Schema::table('users', function (Blueprint $table) {
    $table->unsignedBigInteger('customer_number')->unique()->nullable()->after('id');
});

\Illuminate\Support\Facades\DB::statement('ALTER TABLE users MODIFY customer_number INT AUTO_INCREMENT');
```

This is ugly for two reasons: firstly, you have to do the `ALTER TABLE` outside of the migration function to ensure it runs in the correct order. Secondly, this will cause significant issues if you change the underlying database; the above code will only work for MySQL databases, and would throw errors for any other type. This breaks the purpose of having abstraction layers in the first place.

# The Solution

This pull requests allows for backwards compatibility, still accepting the same behaviour for passing any of the integer-based columns a `true` or `false` value for `$autoIncrements`. However, it additionally adds the ability to pass in a string to indicate the key type. For the databases that support it (all grammars except sqlite), you can pass `unique` instead.

Eg.

```php
Schema::table('users', function (Blueprint $table) {
     $table->uuid('id')->primary();
     $table->string('name');
     $table->integer('customer_number', 'unique');
});

// or

Schema::table('users', function (Blueprint $table) {
     $table->uuid('id')->primary();
     $table->string('name');
     $table->integer('customer_number')->autoIncrements('unique');
});
```

# The Constraints
The primary constraint for this change is keeping it backwards compatible; this is not a big enough change to break BWC, and so this is the most eloquent way I could come up with to allow this functionality.

I considered adding an additional property, however this didn't make sense for when you didn't want a column to be auto incrementing. Instead, I have expanded the meaning of the increment parameter to include the key type.

If anyone has any suggestion of how to achieve this better, then I am open to changing the approach.

It is also worth noting that this raises another issue: sqlite. Sqlite is the only supported grammar that does _not_ support non-primary auto-increments (at least, from what I can see from the documentation).

### CODE REVIEW

**Comment by taylorotwell** [comment]
Why not leave the auto-incrementing ID as your primary key and have a secondary unique key on a `uuid` field?

**Comment by olliescase** [comment]
I think that is definitely a valid, and, admittedly, good way of doing it.

However, I'm not sure that satisfies a reason to prohibit it. Following is a brain-dump several reasons in which I can think why someone might wish to do it. The truth of the matter is that none of these are absolutely game-breaking, but it feels unnecessarily limiting in an otherwise intuitive framework.

For starters, this provides problems for a development flow. You might well introduce auto-numbers as an afterthought in a down-the-line feature. It feels wrong to change your primary key on the table to a different column, and it could well cause large restructuring.

Secondly, primary keys and unique keys are _almost_ the same, and _can_ be the same, but not necessarily _actually_ the same. By default, primary keys are clustered indexes, vs auto-increments being non-clustered. This could potentially have a performance impact if your primary key is different to the key in which you use to primarily identify the record. This is mitigable, but causes more friction in the development flow.

Thirdly, there is a potentially a case in which you would only want to auto-number new records, not legacy ones. By using the auto-number as a primary key, this is not possible.

Fourthly, I think syntactically it causes confusion: a lot of people will assume that the primary key is the column being matched on. Particularly if you are looking at the database, it feels wrong to access throughout the entire codebase the record via a non-primary ID, only to see that the primary ID is _not_ what is being used.

And personally, I like to keep my identifying column called `id` (whether its a uuid or autoincrement ID). To me this is easiest for everyone else reading the code. And there's something that feels off about the following structure:
```
id - char(36) - unique
invoice_number - unsigned integer - primary
name - varchar
...
```

It feels dirty to not have the `id` column be the primary key, and to make another column the uuid feels like it causes unnecessary confusion (and model configuration) for other developers.

I apologise my response is a bit rambly, and very long winded, but I hope this gives some insight into why I think auto-numbers shouldn't be limited to primary keys only, particularly as its supported by most databases.

**Comment by olliescase** [comment]
TLDR; I think that if you wish to not have non-primary auto-increment columns then that should be a per-team policy, not a framework-enforced policy.
